connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java

```diff
@@ -824,10 +830,75 @@ private void transitionTo(Object connectorOrTask, TargetState state, ClassLoader
         }
     }
 
+    ConnectorStatusMetricsGroup connectorStatusMetricsGroup() {
+        return connectorStatusMetricsGroup;
+    }
+
     WorkerMetricsGroup workerMetricsGroup() {
         return workerMetricsGroup;
     }
 
+    class ConnectorStatusMetricsGroup {
+        private ConnectMetrics connectMetrics;
+        private ConnectMetricsRegistry registry;
+        private ConcurrentMap<String, MetricGroup> connectorStatusMetrics = new ConcurrentHashMap<>();
+
+
+        protected ConnectorStatusMetricsGroup(ConnectMetrics connectMetrics) {
+            this.connectMetrics = connectMetrics;
+            this.registry = connectMetrics.registry();
+        }
+
+        protected synchronized void recordTaskAdded(ConnectorTaskId connectorTaskId) {
+            if (connectorStatusMetrics.containsKey(connectorTaskId.connector())) {
+                return;
+            }
+
+            String connName = connectorTaskId.connector();
+
+            MetricGroup metricGroup = connectMetrics.group(registry.workerGroupName(),
+                registry.connectorTagName(), connName);
+
+            metricGroup.addValueMetric(registry.connectorTotalTaskCount,
+                now -> tasks.values()
+                    .stream()
+                    .filter(task -> task.id().connector().equals(connName))
+                    .count());
+            for (Map.Entry<MetricNameTemplate, TaskStatus.State> statusMetric : registry.connectorStatusMetrics
+                .entrySet()) {
+                metricGroup.addValueMetric(statusMetric.getKey(),
+                    now -> tasks.values()
+                        .stream()
+                        .filter(task ->
+                            task.id().connector().equals(connName)
+                            && herder.taskStatus(task.id())
+                            .state()
+                            .equalsIgnoreCase(statusMetric.getValue().toString())
+                        )
+                        .count());
```

Comment:
Similarly, it would be easier to test if we had a method return the lambda function:
```
protected LiteralSupplier<Integer> taskStatusCounter(String connName, TaskStatus.State state) {
  return (now) -> tasks.values()
                       .stream()
                       .filter(task -> 
                               task.id().connector().equals(connName)
                               && herder.taskStatus(task.id())
                                        .state()
                                        .equalsIgnoreCase(state.toString())
                       )
                       .count());
}
```
so the highlighted lines can be:
```
            metricGroup.addValueMetric(
                statusMetric.getKey(), 
                taskStatusCounter(connName, statusMetric.getValue()));
```